proposals/kapp-controller/004-health-status-reporting/README.md

```diff
@@ -0,0 +1,76 @@
+
+We intend to extend the existing App API by adding a new status condition to expose the system's health. To do so, the following needs to be implemented:
+
+1. The controller reconciling the App CR needs to dynamically set up watches for the resources being deployed by the package. 
+2. Introduce a `Healthy` condition in App CR's `status` [field][app_cr_status].
+
+This field would be set to:
+1. `True`: if all the installed resources are healthy.
+2. `False`: if there is any error during the `fetch`, `install`, `deploy` step or if any of the resources are unhelathy.
+
+The proposed set to criteria to determine if a resource is healthy or not is described below:
+
+1. A Pod resource will be healthy if/when: 
+- `.status.Phase` is Running or Completed.
+- If `.status.Phase` is Running, "Available" status condition is True.
+
+2. A ReplicationController resource will be healthy if/when:
+- `.spec.replicas` matches `.status.availableReplicas`
+- `ReplicaFailure` type condition is not present in the status.
+
+3. A ReplicaSet resource will be healthy if/when:
+- `.spec.replicas` matches `.status.availableReplicas`.
+- `ReplicaFailure` type condition is not present in the status.
+
+3. A Deployment resource will be healthy if/when:
+- `.spec.replicas` matches `.status.availableReplicas`.
+- `Available` type condition is true.
+
+4. A StatefulSet resource will be healthy if/when:
+- `.spec.replicas` matches `.status.availableReplicas`.
+- A DaemonSet resource will be healthy if/when:
+`.status.desiredNumberScheduled` matches `.status.numberAvailable`
+
+5. An APIService resource will be healthy if/when: 
+- `Available` type condition in status is true.
+
+6. A CustomResourceDefinition resource will be healthy if/when:
+- `StoredVersions` has the expected API version for the CRD.
+
+7. All other unspecified resources will be considered healthy.
+
+If any of the watched resource is unhealthy, the `Message` field of the healthy condition will have the statuses of the unhealthy resources ordered lexicographically. 
+
+Since the resources deployed by the App reconciler have informers created for them, any change in the resource state will trigger a reconcile that in turn will re-evaluate the health of all resources. 
+
+#### Use Case: Monitoring the state of resources
+
+Kapp currently has the `inspect` command which lists the resources deployed and their current statuses. The output of the command is also printed out as a part of App's status if enabled through `rawOptions` while creating the CR. 
+
+Though this command provides information about the resources created by the respective App CR, it does so by sending API requests during the reconciliation. Instead, using informers provides additional advantages of having real-time updates, efficient resource utilization and reduced load on API server.
+
+### Other Approaches:
+
+## Open Questions:
+
+1. Can using informers to watch resources increase cache size, potentially impacting the performance?
```

Comment:
I think this will be the case since the same kapp-controller can be in charge of hundreds of apps, and if we do this for all the apps, we might end up getting informers for every resource in the cluster.

Reply:
We can have this as an optional feature to start, similar to how `inspect` is right now?

Reply:
If this proposal is implemented, health would ultimately be determined by evaluating only the following kinds:
- Pod
- ReplicationController
- ReplicaSet
- Deployment
- StatefulSet
- APIService
- CustomResourceDefinition

Which would mean the additional overhead is at most 6 more informers with label selectors limiting the cache contents to just what kapp-controller is managing. We don't have to have informers for the App resources that do not contribute to the health condition, right?

Reply:
Yes, it'd only need to be a subset of all APIs

Reply:
Based on the discussion in the community meeting today - 

The two use-cases for setting up informers to watch resources are:
1. Health monitoring and aggregating status.
2. Triggering a reconcile if any resource is unhealthy. 

- From OLM's end, the use case we want to fulfil is (1).  
- (2) is something that can cause performance issues in terms of continuously reconciling for any unhealthy resources even if we have informers set up for limited number of GVK's (especially on clusters where Kapp-ctrl is managing thousands of App CRs).

If (2) is not to be addressed, to maintain modularity in terms of controller's functionality, @joaopapereira suggested we explore having a separate controller to monitor the health status of individual resources which can be optionally enabled.